persist/sqlite/wallet.go

```diff
@@ -434,10 +384,142 @@ WHERE wa.wallet_id=$1 AND sa.sia_address=$2 LIMIT 1`
 	return
 }
 
+func scanSiacoinElement(s scanner) (se types.SiacoinElement, err error) {
+	err = s.Scan(decode(&se.ID), decode(&se.SiacoinOutput.Value), decodeSlice(&se.MerkleProof), &se.LeafIndex, &se.MaturityHeight, decode(&se.SiacoinOutput.Address))
+	return
+}
+
+func scanSiafundElement(s scanner) (se types.SiafundElement, err error) {
+	err = s.Scan(decode(&se.ID), &se.LeafIndex, decodeSlice(&se.MerkleProof), &se.SiafundOutput.Value, decode(&se.ClaimStart), decode(&se.SiafundOutput.Address))
+	return
+}
+
+func insertAddress(tx *txn, addr types.Address) (id int64, err error) {
+	const query = `INSERT INTO sia_addresses (sia_address, siacoin_balance, immature_siacoin_balance, siafund_balance) 
+VALUES ($1, $2, $3, 0) ON CONFLICT (sia_address) DO UPDATE SET sia_address=EXCLUDED.sia_address 
+RETURNING id`
+
+	err = tx.QueryRow(query, encode(addr), encode(types.ZeroCurrency), encode(types.ZeroCurrency)).Scan(&id)
+	return
+}
+
+func fillElementProofs(tx *txn, indices []uint64) (proofs [][]types.Hash256, _ error) {
```

Comment:
It's a bit of a shame that the implementations of the full mode and the partial one are so fragmented persistence wise. On the one hand we have the table of elements with a proof stored directly in their table and then we have the state tree table which replaces the proofs from the element table.

Would it be possible to remove the proof from the element table completely and always use the state tree table? But depending on the mode we either add the whole tree to the table or just the branches we are interested in?

Reply:
We could, but I'm not sure it's worth the hassle. We'd also have to decide whether to clean up deleted state tree nodes with a reference counter or FK or leave them in the database. However, at that point, we might as well store the whole tree anyway. It's not *that* big.